docs/design/sharding_controller.md

```diff
@@ -0,0 +1,309 @@
+    F --> H[Queue Sync Event]
+    G --> I[Done]
+    H --> I
+```
+
+
+### 4.3 Sharding Manager
+The core logic that implements node assignment strategies:
+
+```go
+type SchedulerConfig struct {
+    Name          string
+    Type          string // "volcano" or "agent"
+    ShardStrategy ShardStrategy
+}
+
+type ShardStrategy struct {
+    CPUUtilizationRange struct {
+        Min float64
+        Max float64
+    }
+    PreferWarmupNodes bool
+    MinNodes          int
+    MaxNodes          int
+}
+```
+
+**Assignment Algorithm**:
+1. **Filtering**: For each scheduler, filter nodes matching CPU utilization range
+2. **Scoring**: Score nodes based on warmup preference and utilization
+3. **Selection**: Select top nodes within min/max constraints
+4. **Mutual Exclusivity**: Ensure no node is assigned to multiple schedulers
+
+### 4.4 Event Processing System
+A two-queue system for efficient event handling:
+
+1. **Main Queue**: Processes NodeShard updates and sync events
+2. **Node Event Queue**: Handles node and pod events for metrics updates
+
+```go
+// Event processing workflow
+func (sc *ShardingController) nodeEventWorker() {
+    for sc.processNodeEvent() {
+        // Process node events with rate limiting
+    }
+}
+
+func (sc *ShardingController) worker() {
+    for sc.processNextItem() {
+        // Process shard sync events
+    }
+}
+```
+
+## 5. Event Flow
+
+### 5.1 Node/Pod Events
+1. Pod added/updated/deleted on a node
+2. Controller detects event and updates node metrics
+3. If utilization changes significantly (>10%), trigger shard sync
+4. Sharding Manager recalculates assignments
+5. NodeShard CRDs are updated with new assignments
+
+### 5.2 Periodic Sync
+1. Controller triggers periodic sync (default: 60s)
+2. Sharding Manager recalculates assignments for all nodes
+3. Only update NodeShard CRDs if assignments change significantly
+4. Minimize unnecessary updates to reduce API server load
+
+### 5.3 Initial Startup
+1. Start informers and wait for cache sync
+2. Initialize node metrics for all nodes
+3. Calculate initial assignments
+4. Create NodeShard CRDs for each scheduler
+
+## 6. Node Metrics Calculation
+
+### 6.1 Data Sources
+Instead of external metrics servers, we use:
+- **Pod Resource Requests**: From pod specifications
+- **Node Capacity**: From node status
+- **Node Labels**: For warmup node identification
+
+### 6.2 Calculation Method
+```go
+// Calculate CPU utilization
+totalCPURequest := sum(pod.spec.containers.resources.requests.cpu)
+nodeCPUCapacity := node.status.capacity.cpu
+cpuUtilization := totalCPURequest / nodeCPUCapacity
+```
+
+### 6.3 Update Strategy
+- **Event-Driven**: Update metrics when pod events occur
+- **Periodic Refresh**: Full metrics refresh every 30 seconds
+- **Significant Change Detection**: Only trigger shard sync when utilization changes >10%
+
+## 7. Shard Assignment Algorithm
+
+### 7.1 Hard Filtering Approach
+Unlike complex scoring systems, we use hard filtering:
```

Comment:
This is too simplistic, could you change it to a flowchart? Thanks. And this shouldn't be called hard filtering, it should be called utilization-based filtering, right? Also, why does the code only consider the CPU? Furthermore, it doesn't explain here current sharding logic is based on the interval [min, max].